### PR TITLE
docs: document label expressions

### PR DESCRIPTION
**Pull request by nklaassen**
This commit adds documentation for the label expressions feature described in RFD 116.

### CODE REVIEW

File: docs/pages/access-controls/reference.mdx [hunk 1]
```diff
@@ -12,7 +12,7 @@ controls (RBAC) in your Teleport cluster.
 A Teleport `role` works by having two lists of rules: `allow` rules and `deny` rules.
 When declaring access rules, keep in mind the following:
 
-- Everything is denied by default.
+- Nothing is allowed by default.
```
Comment (nklaassen) [inline docs/pages/access-controls/reference.mdx:15]: I think this is an important distinction, when a role explicitly denies something that means that no other role can allow access to it, even through an access request.

File: docs/pages/includes/role-spec.mdx [hunk 2]
```diff
@@ -120,17 +120,19 @@ spec:
...
       # a list of alternative options:
       'region': ['us-west-1', 'eu-central-1']
-      # regular expressions start with ^ and end with $
-      # Teleport uses golang regexp syntax.
-      # of the list example above can be expressed as:
+      # Regular expressions start with ^ and end with $.
+      # Teleport uses Go's regular expression syntax:
+      # https://github.com/google/re2/wiki/Syntax
+      # The list example above can be expressed as:
+      # 'region': '^us-west-1|eu-central-1$'
```
Comment (nklaassen) [inline docs/pages/includes/role-spec.mdx:135]: I copied this text from the existing comment on `app_labels`, it seems much better

File: docs/pages/reference/predicate-language.mdx [hunk 3]
```diff
@@ -68,3 +70,56 @@ The language also supports the following functions:
...
+
+The language supports the following functions:
+
+[//]: # (comment: some of the code snippets below are strategically broken up to allow line breaks)
+
+| Syntax | Return type | Description | Example |
+|--------|-------------|-------------|---------|
+| `contains(list, item)` | Boolean | Returns true if `list` contains an exact match for `item` | `contains(user.spec.traits[teams], labels["team"])` |
+| `regexp.match(list, re)` | Boolean | Returns true if `list` contains a match for `re` | `regexp.match(labels["team"], "dev-team-\d+$")` |
+| `regexp.replace(list,` `re, replacement)` | `[]string` | Replaces all matches of `re` with replacement for all items in `list` | `contains(regexp.replace(user.spec.traits["allowed-env"],` `"^env-(.*)$", "$1"), labels["env"])`
```
Comment (nklaassen) [inline docs/pages/reference/predicate-language.mdx:105]: I intentionally added some spaces in a few of these code snippets to allow line breaks in acceptable positions

File: docs/pages/access-controls/reference.mdx [hunk 4]
```diff
@@ -228,6 +231,57 @@ spec:
       'environment': '^test|staging$'
 ```
 
+### Label expressions
+
+<Admonition type="warning">
+Label expressions are available starting in Teleport version `13.1.1`.
+All components of your Teleport cluster must be upgraded to version `13.1.1`
+or newer before you will be able to use label expressions.
+This includes the Auth service and **all** Node services.
```
Comment (ptgott) [inline docs/pages/access-controls/reference.mdx:240]:
```suggestion
This includes the Auth Service and **all** Teleport agents.
```
To match the Core Concepts page

File: docs/pages/reference/predicate-language.mdx [hunk 5]
```diff
@@ -68,3 +70,54 @@ The language also supports the following functions:
...
+| `email.local(list)` | `[]string` | Returns the local part of each email in `list`, or an error if any email fails to parse | `contains(email.local(user.spec.traits["email"]),` `labels["owner"])`
+| `strings.upper(list)` | `[]string` | Converts all items of the list to uppercase | `contains(strings.upper(user.spec.traits["username"]),` `labels["owner"])`
+| `strings.lower(list)` | `[]string` | Converts all items of the list to lowercase | `contains(strings.lower(user.spec.traits["username"]),` `labels["owner"])`
+| `labels_matching(re)` | `[]string` | Returns the aggregate of all label values with keys matching `re`, which can be a glob or a regular expression. | `contains(labels_matching("^project-(team\|label)$"),` `"security")`
```
Comment (ptgott) [inline docs/pages/reference/predicate-language.mdx:107]:
```suggestion
| `labels_matching(re)` | `[]string` | Returns the aggregate of all label values with keys matching `re`, which can be a glob or a regular expression | `contains(labels_matching("^project-(team\|label)$"),` `"security")`
```
For consistent punctuation

**Review by ptgott — APPROVED** [review]
Approved with minor suggestions

**Review by rosstimothy — APPROVED** [review]

**Review by zmb3 — APPROVED** [review]

File: docs/pages/access-controls/reference.mdx [hunk 6]
```diff
@@ -228,6 +231,57 @@ spec:
...
+- perform matching on label keys
+- implement negative matches
+
+The following example role would allow access to any nodes labeled `env=staging`
+or labeled `team=<team>`, where `<team>` is one of the values of the user's
+`teams` trait:
+
+```yaml
+kind: role
+version: v5
```
[the remark this answers is not in the document]
Comment (nklaassen) [reply docs/pages/access-controls/reference.mdx]: They actually work fine all the way back to role v3. Our role versioning is weird, it really just changes the meaning of empty/default values, and we manually block some features in older versions, but I don't think there's any reason to block label expressions

File: docs/pages/access-controls/reference.mdx [hunk 7]
```diff
@@ -228,6 +231,57 @@ spec:
...
+metadata:
+  name: example-role
+spec:
+  allow:
+    node_labels_expression: |
+      labels["env"] == "staging" ||
+      contains(user.spec.traits["teams"], labels["team"])
+```
+
+The `<kind>_labels_expression` fields have the same purpose of the
```
[the remark this answers is not in the document]
Comment (nklaassen) [reply docs/pages/access-controls/reference.mdx:266]: You can specify both if you really want to, added a short blurb to explain that

**Pull request merged**